evap/results/tools.py

```diff
@@ -171,42 +170,101 @@ def _calculate_results_impl(course):
+
 
-    for __, contributor, __, results, __ in calculate_results(course):
-        average_likert = avg([result.average for result in results if result.question.is_likert_question])
-        deviation_likert = avg([result.deviation for result in results if result.question.is_likert_question])
-        average_grade = avg([result.average for result in results if result.question.is_grade_question])
-        deviation_grade = avg([result.deviation for result in results if result.question.is_grade_question])
+def sum_distributions(distributions):
+    distributions = [distribution for distribution in distributions if distribution]
+    if distributions:
+        summed_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+        for distribution in distributions:
+            for answer, value in distribution.items():
+                summed_distribution[answer] += value
+        return summed_distribution
+    return None
 
-        (avg_contribution_likert if contributor else avg_generic_likert).append(average_likert)
-        (dev_contribution_likert if contributor else dev_generic_likert).append(deviation_likert)
-        (avg_contribution_grade if contributor else avg_generic_grade).append(average_grade)
-        (dev_contribution_grade if contributor else dev_generic_grade).append(deviation_grade)
 
-    # the final total grade will be calculated by the following formula (GP = GRADE_PERCENTAGE, CP = CONTRIBUTION_PERCENTAGE):
-    # final_likert = CP * likert_answers_about_persons + (1-CP) * likert_answers_about_courses
-    # final_grade = CP * grade_answers_about_persons + (1-CP) * grade_answers_about_courses
-    # final = GP * final_grade + (1-GP) * final_likert
+def multiply_distribution(distribution, factor):
+    if distribution == None:
+        return None
+    for answer, value in distribution.items():
+        distribution[answer] = factor * value
+    return distribution
 
-    final_likert_avg = mix(avg(avg_contribution_likert), avg(avg_generic_likert), settings.CONTRIBUTION_PERCENTAGE)
-    final_likert_dev = mix(avg(dev_contribution_likert), avg(dev_generic_likert), settings.CONTRIBUTION_PERCENTAGE)
-    final_grade_avg = mix(avg(avg_contribution_grade), avg(avg_generic_grade), settings.CONTRIBUTION_PERCENTAGE)
-    final_grade_dev = mix(avg(dev_contribution_grade), avg(dev_generic_grade), settings.CONTRIBUTION_PERCENTAGE)
 
-    final_avg = mix(final_grade_avg, final_likert_avg, settings.GRADE_PERCENTAGE)
-    final_dev = mix(final_grade_dev, final_likert_dev, settings.GRADE_PERCENTAGE)
+def avg_distribution(distributions, weights=None):
+    if not weights:
+        return normalize_distribution(sum_distributions(distributions))
 
-    return final_avg, final_dev
+    weighted_distributions = []
+    for distribution, weight in zip(distributions, weights):
+        weighted_distributions.append(multiply_distribution(distribution, weight))
+    return normalize_distribution(sum_distributions(weighted_distributions))
```

Comment:
i would make this here as analogous to the `avg` as possible. that would include skipping `None`s, and then you could remove the None handling from normalize, sum and multiply.

... actually, that would also include inlining sum and multiply, i think. although it's probably not really inlining, i think most of the code will just disappear. could you try that?

Reply:
ah, so there's more stuff with avg. so then disregard the previous comment and i will right now review everything myself because i can't continue without getting an overview again :)